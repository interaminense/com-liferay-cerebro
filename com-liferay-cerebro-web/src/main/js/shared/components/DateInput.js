import autobind from 'autobind-decorator';
import moment from 'moment';
import {noop} from 'lodash';

import Button from './Button';
import Component, {Config} from 'metal-jsx';
import DatePicker from './date-picker';
import Icon from './Icon';
import Input from './Input';
import Overlay from './Overlay';
import {FORMAT} from 'shared/util/date';

class DateInput extends Component {
	@autobind
	handleClick(event) {
		event.preventDefault();

		this.state.active = true;
	}

	@autobind
	handleDateSelect(value) {
		const {onChange, format} = this.props;

		onChange(value.format(format));

		this.state.active = false;
	}

	@autobind
	handleOutsideClick() {
		this.state.active = false;
	}

	render() {
		const {props: {format, value}, state: {active}} = this;

		const date = moment(value, format);

		return (
			<Overlay
				active={active}
				containerClass="date-input-root"
				onOutsideClick={this.handleOutsideClick}
			>
				<Input.Group inset>
					<Input.GroupInput>
						<Input
							{...this.otherProps()}
							onClick={this.handleClick}
							value={value}
						/>
					</Input.GroupInput>

					<Input.Inset>
						<Button display="unstyled" onClick={this.handleClick}>
							<Icon symbol="calendar" />
						</Button>
					</Input.Inset>
				</Input.Group>

				<DatePicker
					date={date.isValid() ? date : null}
					onSelect={this.handleDateSelect}
				/>
			</Overlay>
		);
	}
}

DateInput.STATE = {
	active: Config.bool().value(false)
};

DateInput.PROPS = {
	format: Config.string().value(FORMAT),
	onChange: Config.func().value(noop),
	value: Config.string()
};

export default DateInput;