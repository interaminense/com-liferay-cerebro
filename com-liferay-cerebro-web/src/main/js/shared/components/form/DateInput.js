import autobind from 'autobind-decorator';
import Component, {Config} from './Component';
import getCN from 'classnames';
import {isNumber, uniqueId} from 'lodash';

import DateInput from '../DateInput';
import Label from './Label';
import validate from '../../util/form-validators';

class FormDateInput extends Component {
	created() {
		const {initialValue} = this.props;

		if (initialValue) {
			this.state.value = initialValue;
		}

		this._id = uniqueId('FormDateInput');
	}

	@autobind
	handleBlur() {
		this.validateFormElement();
	}

	@autobind
	handleChange(value) {
		this.setValue(value);
	}

	@autobind
	handleInput(event) {
		this.setValue(event.target.value);
	}

	setValue(value) {
		this.state.value = value;

		if (this.state.show) {
			this.validateFormElement();
		}
	}

	validateFormElement(showErrors = true) {
		const {validator} = this.props;

		let retVal = true;

		if (validator) {
			const result = validate(this.state.value, {
				...validator,
				date: true
			}).find(item => !item.valid);

			const valid = result ? result.valid : true;

			this.setState({
				message: result ? result.message : '',
				show: showErrors,
				valid
			});

			retVal = valid;
		}

		return retVal;
	}

	syncInitialValue(newVal) {
		this.state.value = newVal;
	}

	render() {
		const {
			props: {info, inline, label, width},
			state: {message, show, valid, value}
		} = this;

		const classes = getCN('form-date-input-root', {
			'form-inline-group': inline,
			'has-danger': !valid && show,
			'has-success': valid && show
		});

		const style = isNumber(width)
			? {'flex-basis': `${width}%`, 'flex-grow': 0}
			: undefined;

		return (
			<div class={classes} style={style}>
				{label && (
					<Label
						for={this._id}
						info={info}
						required={this.isRequired()}
					>
						{label}
					</Label>
				)}

				<DateInput
					{...this.otherProps()}
					id={this._id}
					onBlur={this.handleBlur}
					onChange={this.handleChange}
					onInput={this.handleInput}
					value={value}
				/>

				{show && <div class="help-block">{message}</div>}
			</div>
		);
	}
}

FormDateInput.PROPS = {
	info: Config.string(),
	initialValue: Config.string().value(''),
	inline: Config.bool().value(false),
	label: Config.string(),
	validator: Config.object()
};

FormDateInput.STATE = {
	message: Config.value(''),
	show: Config.value(false),
	valid: Config.value(true),
	value: Config.value('')
};

export default FormDateInput;