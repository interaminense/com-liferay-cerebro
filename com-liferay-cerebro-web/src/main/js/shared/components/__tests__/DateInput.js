import moment from 'moment';

import DateInput from '../DateInput';
import {renderWithStore} from 'test/mock-store';

describe('DateInput', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = renderWithStore(DateInput);

		expect(component).toMatchSnapshot();
	});

	it('should call onChange', () => {
		const onChange = jest.fn();

		component = renderWithStore(DateInput, {onChange});

		expect(onChange).not.toHaveBeenCalled();

		component.handleDateSelect(moment(0));

		expect(onChange).toHaveBeenCalled();
	});
});