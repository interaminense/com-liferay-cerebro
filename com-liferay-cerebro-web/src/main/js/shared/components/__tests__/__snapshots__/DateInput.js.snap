// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput should render 1`] = `
<div class="input-group input-group-inset" id="overlay1">
  <div class="input-group-input">
    <input class="input-root form-control" type="text">
  </div>
  <span class="input-group-inset-item">
    <button class="button-root btn btn-unstyled" type="button">
      <svg class="icon-root lexicon-icon lexicon-icon-calendar">
        <use xlink:href="/images/lexicon/icons.svg#calendar"></use>
      </svg>
    </button>
  </span>
</div>
`;
